### PR TITLE
convenience functions for client-go transport

### PR DESCRIPTION
**Pull request by p0lyn0mial**
this PR adds a few convenience functions that I need in https://github.com/openshift/kubernetes/pull/603

### CODE REVIEW

File: pkg/config/client/transport.go [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+}
+
+func (t *preferredHostRT) RoundTrip(r *http.Request) (*http.Response, error) {
+	preferredHost := t.preferredHostFn()
+
+	if len(preferredHost) == 0 {
+		return t.baseRT.RoundTrip(r)
+	}
+
+	r.Host = preferredHost
```
Comment (p0lyn0mial) [inline pkg/config/client/transport.go:35]: copying here means that we won't see it in the logs, is that what we want?

File: pkg/config/client/transport.go [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+	"net/http"
+
+	"k8s.io/client-go/rest"
+)
+
+// AnonymousClientConfigWithWrapTransport returns a copy of the given config with all user credentials (cert/key, bearer token, and username/password) and custom transports (Transport) removed.
+// This function preserves WrapTransport for clients that care about custom HTTP behavior.
+func AnonymousClientConfigWithWrapTransport(config *rest.Config) *rest.Config {
+	newConfig := rest.AnonymousClientConfig(config)
+	newConfig.WrapTransport = config.WrapTransport
```
Comment (soltysh) [inline pkg/config/client/transport.go:13]: I literally asked for that upstream :wink:

**Review by soltysh — APPROVED** [review]
/ltm

File: pkg/config/client/transport.go (hunk 2 above)
Comment (p0lyn0mial) [reply pkg/config/client/transport.go:13]: heh :)

**Comment by p0lyn0mial** [comment]
had to fix the `verify` job

File: pkg/config/client/transport.go [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+func (t *preferredHostRT) RoundTrip(r *http.Request) (*http.Response, error) {
+	preferredHost := t.preferredHostFn()
+
+	if len(preferredHost) == 0 {
+		return t.baseRT.RoundTrip(r)
+	}
+
+	r.Host = preferredHost
+	r.URL.Host = preferredHost
```
Comment (sttts) [inline pkg/config/client/transport.go:38]: don't we have to keep the tls server name in order to verify the serving cert successfully? Or do we have localhost certs?

Comment (p0lyn0mial) [reply pkg/config/client/transport.go:38]: yes, we do have CA that knows how to verify the serving cert for localhost at least for KCM.

Comment (sttts) [reply pkg/config/client/transport.go:38]: That way you could also use a node IP if necessary.

Comment (p0lyn0mial) [reply pkg/config/client/transport.go:38]: we decided to set the `ServerName` at https://github.com/openshift/kubernetes/pull/603/commits/e5a8075fb932d3f91f8134d941d280a21149ded1#diff-d073b3d85bfc4877d354264d569dd3e19d544d1775a67f1650d661299869aa7aR448

File: pkg/config/client/transport.go [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+func AnonymousClientConfigWithWrapTransport(config *rest.Config) *rest.Config {
+	newConfig := rest.AnonymousClientConfig(config)
+	newConfig.WrapTransport = config.WrapTransport
+	return newConfig
+}
+
+// OverwriteHostInTLSServerName extract the hostname from the config.Host and sets it in config.ServerName
+// the ServerName is passed to the server for SNI and is used in the client to check server certificates.
+func OverwriteHostInTLSServerName(config *rest.Config) error {
+	u, err := url.Parse(config.Host)
```
Comment (p0lyn0mial) [inline pkg/config/client/transport.go:21]: `config.Host` is always scheme:host:port, where port is optional

**Pull request merged**